### PR TITLE
add `method` and `weight` feature to (region) aggregation

### LINKED ISSUE #299: Suggest to change default behaviour of `aggegate_region()` for components
## Introduction

The current implementation of [aggegate_region()](https://pyam-iamc.readthedocs.io/en/stable/api.html#pyam.IamDataFrame.aggregate_region) has an option to add `components` (other variables), which are only defined at the (aggregate) `region` level. Per default, the function uses all sub-categories of `variable` that are only defined at the `region` level (not in any of the `subregions`).

## Problem

The current implementation is very slow when iterating over many variables, and it might cause an issue when the sequence of aggregation matters on the outcome.

## Suggested solution

Allow `variable` to be a list of variables rather than only one (string), so that the region-aggregation can be performed in bulk. But this will cause problems with the automated detection of `components`. Also, passing the components for individual variables in the list would be cumbersome... (probably require a dictionary, what if that is not complete, ...).

Therefore, I suggest the following API change:
 - `variable` is a string or a list of strings, `components` is False (default): do not add any components, perform region aggregation on single variable or entire list (per variable)
 - `variable` is a string, `components` is True: automatically detect sub-categories (current behaviour)
 - `variable` is a string, `components` is a list: use `components` for the region-only-level values to be added
 - if `variable` is a list, `components` must be False

## Required changes in existing scripts

To continue current behaviour, existing scripts would need to add `components=True`.

### PR DESCRIPTION
**Pull request by danielhuppmann**
# Please confirm that this PR has done the following:

- [x] Tests Added
- [x] Documentation Added
- [x] Description in RELEASE_NOTES.md Added

# Description of PR

This PR extends the functions `aggregate()`, `aggregate_region()` and the respective check-versions by a `method` and `weight` options (weights only for region aggregation). The functionality is based on the branch [peterkolp:region_aggregation_mip_feature](https://github.com/peterkolp/pyam/tree/region_aggregation_mip_feature).

The PR also changes the default of `aggregate_region()` regarding the treatment of components at the region-level as discussed in #299.
> **tl; dr**
> Before, the default behaviour of `aggregate_region()` was to add all sub-categories of `variable` that were not present in any subregion, and deactivating this addition required to set `components=[]`
> Now, the default is `components=False` (not adding sub-categories at the region-level) and `components=True` activates the automatic detection of sub-categories.
> Setting a specific list of variables to be added via `components=<list>` is unchanged.

closes #299

### CODE REVIEW

**Comment by danielhuppmann** [comment]
This PR is directed into `IAMconsortium:tests/cleanup` to highlight the differences of the aggregation-features to that prep-work - I'll rebase after a first round of review and merging of #304.

Once this PR is merged, I'll implemented the feature that `variable` can be a list of strings rather than only one string to speed up the processing.

@byersiiasa @gidden @znicholls @peterkolp @zikolach @volker-krey - any volunteers for a review?

File: pyam/core.py [hunk 1]
```diff
@@ -1444,11 +1484,45 @@ def _meta_idx(data):
...
+def _aggregate(df, by, method=np.sum):
     """Aggregate `df` by specified column(s), return indexed `pd.Series`"""
     by = [by] if isstr(by) else by
     cols = [c for c in list(df.columns) if c not in ['value'] + by]
-    return df.groupby(cols).sum()['value']
+    # pick aggregator func (default: sum)
+    return df.groupby(cols)['value'].agg(_get_method_func(method))
+
+
+def _aggregate_weight(df, weight, method):
```
Comment (byersiiasa) [inline pyam/core.py]: if this is for weighted average, then surely method is always sum/np.sum? or is this just meant as a placeholder for future methods?

Consider also instead using `np.nansum` such that if a value is missing / nan, the calculation won't return `nan`

File: pyam/core.py [hunk 2]
```diff
@@ -1444,11 +1484,45 @@ def _meta_idx(data):
...
+        raise ValueError('only method `np.sum` allowed for weighted average')
+
+    _data = _get_value_col(df, YEAR_IDX)
+    _weight = _get_value_col(weight, YEAR_IDX)
+
+    if not _data.index.equals(_weight.index):
+        raise ValueError('inconsistent index between variable and weight')
+
+    cols = META_IDX + ['year']
+    return (_data * _weight).groupby(cols).sum() / _weight.groupby(cols).sum()
```
Comment (byersiiasa) [inline pyam/core.py]: would you put `.agg(method)` here instead of `.sum()`

Comment (byersiiasa) [reply pyam/core.py]: do you also want to limit inputs to the `KNOWN_FUNCS` ?

File: pyam/core.py (hunk 1 above)
Comment (danielhuppmann) [reply pyam/core.py]: 1) The `method` arg is partly anticipating more supported functions in the future, partly for checking that the chosen method is indeed sum in that function (moving some not-so important stuff to lower-level functions, also for reusing `_aggregate_weight()` in other functions in the future and not needing to implement that check in multiple higher-level functions.

2) The `data` table cannot have `nan` (they are removed at initialisation), so this point is moot. If a data value or the weight value is missing (i.e., inconsistent series index's), a `ValueError` is raised.

File: pyam/core.py (hunk 2 above)
Comment (danielhuppmann) [reply pyam/core.py]: only summation is allowed (for the time being) anyway, so imho no need to be more complicated...

**Comment by gidden** [comment]
Restarted CI (mac py36 only)

File: pyam/core.py (hunk 2 above)
Comment (gidden) [reply pyam/core.py]: Agree that we don't need anything more complicated here, but maybe worth putting in some comments about what would need to be changed in the future if more than sum is supported?

File: pyam/utils.py [hunk 3]
```diff
@@ -33,6 +33,8 @@
                           + ['{}{}'.format(i, j) for i, j in itertools.product(
                               string.ascii_uppercase, string.ascii_uppercase)]))
 
+KNOWN_FUNCS = {'min': np.min, 'max': np.max, 'avg': np.mean, 'sum': np.sum}
```
Comment (gidden) [inline pyam/utils.py]: maybe add a `'mean'` synonym as well?

File: tests/test_feature_aggregate.py [hunk 4]
```diff
@@ -8,6 +8,91 @@
...
+
+    # use other method (max) both as string and passing the function
+    idx = ['model', 'scenario', 'unit', 'year']
+    exp = pd.DataFrame([
+        ['model_a', 'scen_a', 'USD/tCO2', 2005, 10.0],
+        ['model_a', 'scen_a', 'USD/tCO2', 2010, 30.0]
+    ],
+        columns=idx + ['value']
+    ).set_index(idx).value
+    obs = df.aggregate_region('Price|Carbon', method='max')
```
Comment (gidden) [inline tests/test_feature_aggregate.py]: Maybe I don't see it, but is there a test that explicitly tests the `weight` option?

File: pyam/utils.py (hunk 3 above)
Comment (danielhuppmann) [reply pyam/utils.py]: done

File: tests/test_feature_aggregate.py (hunk 4 above)
Comment (danielhuppmann) [reply tests/test_feature_aggregate.py]: We have one (and the line above it tests that not adding the weight yields an unexpected answer)

`assert df.check_aggregate_region('Price|Carbon', weight=v) is None`

**Comment by danielhuppmann** [comment]
Thanks @gidden for the comments! Implemented the one minor change & answered the question about the test inline.

About the "how to support other methods that sum with weights" (seems I can't answer inline) - it's not clear to me what a "weighted max" (or similar) would be expected to do, so no idea what to write as a comment. Happy to discuss implementation when an actual use case comes up.
